sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,373 @@
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+We introduce a new `native_fixed_size` ABI which combines elements of the
+existing `native` and `fixed_size` ABIs defined in Parallelism TS2;
+implementations are expected to use the most efficient representation
+of a `simd<T, N>` type (e.g. using multiple registers rather than an array).
+We expect implementations of this ABI to build on Clang's `ext_vector_type`,
+rather than specifying the register types directly.
+
+The value of `N` for each `simd` argument to a SIMD function must be the same,
+and represents the sub-group size of the calling kernel.  If a function accepts
+`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
+kernel and the implementation must issue a diagnostic in this case.
+
+The example below shows a simple SIMD function that scales all elements of
+an 8-wide SIMD type `x` by a scalar value `n`.
+
+[source, c++]
+----
+sycl::ext::oneapi::experimental::simd<float, 8> scale(sycl::ext::oneapi::experimental::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+----
+
+=== Invoking a SIMD Function
+
+The `invoke_simd` free function invokes a SIMD function using all work-items
+in a `sub_group`.  The `invoke_simd` interface marshals data between the
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+template <typename Callable, typename...T>
+auto invoke_simd(sub_group sg, Callable&& f, T... args);
+
+} // namespace experimental
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+The `Callable` passed to `invoke_simd` must be a C++ callable that can be
+invoked with the same number of arguments specified in the `args` parameter
+pack. `Callable` may be a function object, a lambda,
+or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
+`Callable` must be an immutable callable with the same type and state for all
+work-items in the sub-group, otherwise behavior is undefined.
+
+NOTE: When using a lambda `Callable`, it is very easy for a user to
+unintentionally capture non-uniform variables and violate the requirements
+from the previous paragraph. This form of `invoke_simd` should be used with
+caution.
+
+Each argument to the invoked `Callable` must be a scalar type, a SIMD type,
+or a `std::tuple` of these types. The length of each SIMD type must match the
+sub-group size of the kernel; calling a SIMD function with mismatched lengths
+is an error, and an implementation must issue a diagnostic.
+
+To enable overloading in cases where the `Callable` accepts either a scalar
+type or a SIMD type, we leverage the `sycl::ext::oneapi::experimental::uniform` wrapper
+type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.
+
```

Comment:
one more late nit: it would be good to have an example how overloaded callables should be passed to invoke_simd. E.g. this won't compile - scale1 is overloaded, so `scale1` as a function pointer is ambiguous (thanks to @skozhukh for bringing up). I guess there are general C++ tools for that, but since overloading is mentioned, it would help to see how it is supposed to be done with invoke_simd.

```cpp
float scale1(sycl::ext::oneapi::simd<float, 8> x, float n) {
  return x[2] * n;
}

float scale1(sycl::ext::oneapi::simd<float, 8> x, sycl::ext::oneapi::simd<float, 8> n) {
  return x[2] * n[2];
}

...
  float x = 1;
  float n = 2;
  // invoke SIMD function
  // x values from each work-item are combined into a simd<float, 8>
  auto y = sycl::ext::oneapi::invoke_simd(
      sg, scale1, x, sycl::ext::oneapi::uniform{n});

```

Reply:
... like this:
```cpp
  auto y = sycl::ext::oneapi::invoke_simd<float(sycl::ext::oneapi::simd<float, 8>, float)>(
      sg, scale1, x, sycl::ext::oneapi::uniform{n});
```
?

Reply:
That would be one way to write it.  Anything that causes `scale1` to become a pointer to a specific function will work (see https://en.cppreference.com/w/cpp/language/overloaded_address).

So something like this would work, too:
```c++
auto f = static_cast<float(sycl::ext::oneapi::simd<float, 8>, float)>(scale1);
auto y = sycl::ext::oneapi::invoke_simd(sg, f, x, sycl::ext::oneapi::uniform{n});
```

There's also a neat trick where you can "lift" a function via a lambda, so that it would automatically call the correct version of `scale1` based on the arguments.  There's a good summary of that approach here: https://www.bfilipek.com/2019/03/lambdas-story-part2.html#bonus-lifting-with-lambdas.

This does make me think that the following line from the specification may be wrong:

> After this mapping has taken place, an appropriate variant of `Callable` is selected based on standard overload resolution.

@rolandschulz, what do you think about this?  It certainly looks to me like if the user passes the name of an overloaded function to `invoke_simd`, we have to resolve it _before_ we can call `invoke_simd`.  Is this a bug, or is there a trick I'm missing?

Reply:
I spoke to Roland offline about this.  The specification is correct, but we might want to consider adding a non-normative note and some examples to help people understand how this is expected to work.

The comment about overloading refers to situations where the `Callable` itself represents an overload set (e.g. because it is a functor with multiple call operators).  Your example doesn't fit this case, because what you're passing to `invoke_simd` is the name of an overloaded function (which decays to a function pointer).  It's the user's responsibility to handle that ambiguity.  Note that none of this is really specific to `invoke_simd`; it matches the behavior of functions like `std::invoke` and `std::apply`.

There's no right answer here, but we should adopt a consistent style for training examples.  If the user doesn't want `invoke_simd` to automatically select an overload, the clearest syntax is the one that you proposed:
```c++
  auto y = sycl::ext::oneapi::invoke_simd<float(sycl::ext::oneapi::simd<float, 8>, float)>(sg, scale1, x, sycl::ext::oneapi::uniform{n});
```

If the user does want `invoke_simd` to automatically select an overload, converting the functions into a functor with multiple call operators is the clearest syntax:
```c++
struct scale1 {
  float operator()(sycl::ext::oneapi::simd<float, 8> x, float n) {
    return x[2] * n;
  }

  float operator()(sycl::ext::oneapi::simd<float, 8> x, sycl::ext::oneapi::simd<float, 8> n) {
    return x[2] * n[2];
  }
}
...
  auto y = sycl::ext::oneapi::invoke_simd(sg, scale1(), x, sycl::ext::oneapi::uniform{n});
```
...although note that we now have to pass an instance of `scale1` to `invoke_simd` instead of just the name.

Reply:
Thanks for the explanation, John.

> but we might want to consider adding a non-normative note and some examples to help people understand how this is expected to work

yes, please

Reply:
Added in https://github.com/intel/llvm/pull/3445/commits/b278fb495b8ce763667dc8cacfe1ebbc1c2b8d08.